torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,406 @@
+  }
+
+namespace c10d {
+
+using KeyType = AlgorithmKey;
+using EntryType = std::unique_ptr<AlgorithmEntry>;
+
+namespace {
+
+// Wrap c10d store as Gloo store
+class GlooStore : public ::gloo::rendezvous::Store {
+ public:
+  GlooStore(const std::shared_ptr<::c10d::Store>& store) : store_(store) {}
+
+  void set(const std::string& key, const std::vector<char>& value) override {
+    std::vector<unsigned char> tmp(value.begin(), value.end());
+    store_->set(key, tmp);
+  }
+
+  std::vector<char> get(const std::string& key) override {
+    auto value = store_->get(key);
+    return std::vector<char>(value.begin(), value.end());
+  }
+
+  void wait(const std::vector<std::string>& keys) override {
+    store_->wait(keys, Store::kDefaultTimeout);
+  }
+
+  void wait(
+      const std::vector<std::string>& keys,
+      const std::chrono::milliseconds& timeout) override {
+    store_->wait(keys, timeout);
+  }
+
+ protected:
+  std::shared_ptr<::c10d::Store> store_;
+};
+
+template <typename T>
+const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
+  switch (r) {
+    case ReduceOp::SUM:
+      return ::gloo::ReductionFunction<T>::sum;
+    case ReduceOp::PRODUCT:
+      return ::gloo::ReductionFunction<T>::product;
+    case ReduceOp::MIN:
+      return ::gloo::ReductionFunction<T>::min;
+    case ReduceOp::MAX:
+      return ::gloo::ReductionFunction<T>::max;
+  }
+
+  throw std::runtime_error("Unhandled ReduceOp");
+}
+
+} // namespace
+
+ProcessGroupGloo::WorkGloo::WorkGloo() : completed_(false) {}
+
+ProcessGroupGloo::WorkGloo::~WorkGloo() {}
+
+bool ProcessGroupGloo::WorkGloo::isCompleted() const {
+  return completed_;
+}
+
+bool ProcessGroupGloo::WorkGloo::wait() {
+  std::unique_lock<std::mutex> lock(m_);
+  while (!completed_) {
+    cv_.wait(lock);
+  }
+  return !ex_;
+}
+
+const std::exception& ProcessGroupGloo::WorkGloo::exception() const {
+  return *ex_;
+}
+
+void ProcessGroupGloo::WorkGloo::finish() {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+  }
+  cv_.notify_all();
+}
+
+void ProcessGroupGloo::WorkGloo::finishWithException(
+    const ::gloo::Exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+    ex_ = std::unique_ptr<::gloo::Exception>(new ::gloo::Exception(ex));
+  }
+  cv_.notify_all();
+}
+
+ProcessGroupGloo::Options::Options()
+    : timeout(std::chrono::milliseconds(10 * 1000)), threads(1) {}
+
+ProcessGroupGloo::ProcessGroupGloo(
+    const std::shared_ptr<Store>& store,
+    int rank,
+    int size)
+    : ProcessGroup(rank, size), store_(new GlooStore(store)), stop_(false) {}
+
+ProcessGroupGloo::~ProcessGroupGloo() {
+  // Require this process group to be explicitly shut down prior to being
+  // destructed. This is the easiest way to guarantee clean shutdown
+  // and avoid blocking/throwing in a destructor.
+}
+
+void ProcessGroupGloo::initialize() {
+  Options options;
+  options.timeout = std::chrono::milliseconds(100);
```

Comment:
100ms of timeout seems really strict. Are you sure we don't want to raise this to a few seconds?

Reply:
Good point. This was a leftover from doing testing. But since we can override the options we should set this from the test explicitly.